tests/test_bonded.py

```diff
@@ -128,46 +127,121 @@ def test_rmsd_restraint(self):
         n_particles_a = 25
         n_particles_b = 7
 
-        relative_tolerance_at_precision = {np.float32: 2e-5, np.float64: 1e-9}
 
-        for precision, rtol in relative_tolerance_at_precision.items():
+        for _ in range(100):
+            coords = self.get_random_coords(n_particles_a + n_particles_b, 3)
 
-            for _ in range(100):
-                coords = self.get_random_coords(n_particles_a + n_particles_b, 3)
+            n = coords.shape[0]
+            n_mapped_atoms = 5
 
-                n = coords.shape[0]
-                n_mapped_atoms = 5
+            atom_idxs_a = np.arange(n_particles_a)
+            atom_idxs_b = np.arange(n_particles_b)
 
-                atom_idxs_a = np.arange(n_particles_a)
-                atom_idxs_b = np.arange(n_particles_b)
+            atom_map = np.stack([
+                np.random.randint(0, n_particles_a, n_mapped_atoms, dtype=np.int32),
+                np.random.randint(0, n_particles_b, n_mapped_atoms, dtype=np.int32) + n_particles_a
+            ], axis=1)
 
-                atom_map = np.stack([
-                    np.random.randint(0, n_particles_a, n_mapped_atoms, dtype=np.int32),
-                    np.random.randint(0, n_particles_b, n_mapped_atoms, dtype=np.int32) + n_particles_a
-                ], axis=1)
+            atom_map = atom_map.astype(np.int32)
 
-                atom_map = atom_map.astype(np.int32)
+            params = np.array([], dtype=np.float64)
+            k = 1.35
+            lamb = 0.0
 
-                params = np.array([], dtype=np.float64)
-                k = 1.35
-                lamb = 0.0
+            for precision, rtol, atol in [(np.float64, 1e-6, 1e-6), (np.float32, 1e-4, 1e-6)]:
+
+                ref_u = functools.partial(bonded.rmsd_restraint,
+                    group_a_idxs=atom_map[:, 0],
+                    group_b_idxs=atom_map[:, 1],
+                    k=k
+                )
+
+                test_u = potentials.RMSDRestraint(atom_map, n, k)
+
+                # note the slightly higher than usual atol (1e-6 vs 1e-8)
+                # this is due to fixed point accumulation of energy wipes out
+                # the low magnitude energies as some of test cases have
+                # an infinitesmally small absolute error (on the order of 1e-12)
+                self.compare_forces(
+                    coords,
+                    params,
+                    box,
+                    lamb,
+                    ref_u,
+                    test_u,
+                    rtol,
+                    atol=atol,
+                    precision=precision
+                )
+
+
+    def test_rmsd_restraint_singularity(self):
+        # test the RMSD force by generating random coordinates.
```

Comment:
Could we expand this test's docstring to describe which of [the two identified singularities](https://github.com/proteneer/timemachine/pull/391/files#diff-d0b9851a09ca44d39e9ef2f006a931da3a68103d2a2781370f5157cb05ee4038R131) it is intended to hit, and what the expected behavior is? 

Although it looks to me like this may not quite be testing what you intend, it's not immediately obvious to me how to repair it.

Since this uses `n_A=n_B=3` I think the correlation matrices will always be rank deficient -- and thus `ref_u = 0` based on the check you had added in the previous version. I think this condition is tripped not just for small perturbations of specially chosen `x0, x1` (as currently written), but for most any choice of 6 input coordinates.

(I.e. replacing [L225-L233](https://github.com/proteneer/timemachine/pull/391/files#diff-726b2c071a1c59bd5858e946421e900c75db412d33d202ecd3f285b42bda8e60R225-R233) with `xs = [np.random.randn(6, 3) for _ in range(20)]` doesn't change whether the test passes, and `ref_u` still evaluates to 0 for all `x` in `xs`.)